--- a/include/event2/event_struct.h
+++ b/include/event2/event_struct.h
@@ -94,13 +94,10 @@ struct {								\
 }
 #endif /* !LIST_ENTRY */
 
-#ifndef LIST_HEAD
-#define EVENT_DEFINED_LISTHEAD_
-#define LIST_HEAD(name, type)						\
+#define LIST_HEAD_LEVENT(name, type)						\
 struct name {								\
 	struct type *lh_first;  /* first element */			\
 	}
-#endif /* !LIST_HEAD */
 
 struct event;
 
@@ -163,15 +160,13 @@ TAILQ_HEAD (event_list, event);
 #undef TAILQ_HEAD
 #endif
 
-LIST_HEAD (event_dlist, event); 
+LIST_HEAD_LEVENT (event_dlist, event); 
 
 #ifdef EVENT_DEFINED_LISTENTRY_
 #undef LIST_ENTRY
 #endif
 
-#ifdef EVENT_DEFINED_LISTHEAD_
-#undef LIST_HEAD
-#endif
+#undef LIST_HEAD_LEVENT
 
 #ifdef __cplusplus
 }
