Index: openvpn-2.5.3/src/openvpn/misc.h
===================================================================
--- openvpn-2.5.3.orig/src/openvpn/misc.h
+++ openvpn-2.5.3/src/openvpn/misc.h
@@ -72,7 +72,7 @@ struct user_pass
 #ifdef ENABLE_PKCS11
 #define USER_PASS_LEN 4096
 #else
-#define USER_PASS_LEN 128
+#define USER_PASS_LEN 512
 #endif
     char username[USER_PASS_LEN];
     char password[USER_PASS_LEN];
