--- a/src/init.c
+++ b/src/init.c
@@ -54,7 +54,7 @@
   int tty_set=FALSE;
 
   LOG_ENTER();
-  *port=6400;
+  //*port=6400;
   mdm_init_config(&cfg[0]);
   cfg[0].dte_speed=38400;
   cfg[0].dce_speed=38400;
--- a/src/tcpser.c
+++ b/src/tcpser.c
@@ -54,7 +54,8 @@
 
   modem_count = init(argc, argv, cfg, 64, &port,all_busy,sizeof(all_busy));
 
-  sSocket = ip_init_server_conn(port);
+  if (port != 0)
+  	sSocket = ip_init_server_conn(port);
 
   for(i=0;i<modem_count;i++) {
     if( -1 == pipe(cfg[i].data.mp[0])) {
@@ -69,7 +70,8 @@
       LOG(LOG_FATAL,"Could not open serial port %s",cfg->dce_data.tty);
       exit(-1);
     }
-    cfg[i].line_data.sfd=sSocket;
+    if (port != 0)
+    	cfg[i].line_data.sfd=sSocket;
 
     rc=pthread_create(&thread_id,NULL,*run_bridge,(void *)&cfg[i]);
     if(rc < 0) {
@@ -85,7 +87,7 @@
       FD_SET(cfg[i].data.mp[0][0], &readfs); 
       max_fd=MAX(max_fd,cfg[i].data.mp[0][0]);
     }
-    if(accept_pending==FALSE) {
+    if((accept_pending==FALSE) && (port != 0)) {
       max_fd=MAX(max_fd,sSocket);
       FD_SET(sSocket, &readfs); 
     }
@@ -101,7 +103,7 @@
         }
       }
     }
-    if (FD_ISSET(sSocket,&readfs)) {  // IP traffic
+    if ((port != 0) && FD_ISSET(sSocket,&readfs)) {  // IP traffic
       if(!accept_pending) {
         LOG(LOG_DEBUG,"Incoming connection pending");
         // first try for a modem that is listening.
