--- /src/modem_core.c
+++ /src/modem_core.c
@@ -109,7 +109,7 @@
   cfg->invert_dsr=FALSE;
   cfg->invert_dcd=FALSE;
 
-  cfg->config0[0]='\0';
+  //cfg->config0[0]='\0';
   cfg->config1[0]='\0';
 
   dce_init_config(cfg);
@@ -163,6 +163,16 @@
   return 0;
 }
 
+int mdm_send_init_string(modem_config *cfg) {
+  if(cfg->config0){
+	//printf("sending init string:%s\n",cfg->config0);
+	mdm_write(cfg, cfg->config0, sizeof(cfg->config0));
+  }else{
+	//printf("No init string configured, not sending anything extra\n");
+  }
+  
+  return 0;
+}
 
 void mdm_write_char(modem_config *cfg,unsigned char data) {
   unsigned char str[2];
@@ -635,6 +645,7 @@
 
 int mdm_send_ring(modem_config *cfg) {
   LOG(LOG_DEBUG,"Sending 'RING' to modem");
+  printf("sending ring to modem\n");
   cfg->line_ringing = TRUE;
   mdm_send_response(MDM_RESP_RING,cfg);
   cfg->rings++;

--- /src/bridge.c
+++ /src/bridge.c
@@ -26,7 +26,7 @@
 
 int accept_connection(modem_config* cfg) {
   LOG_ENTER();
-
+  mdm_send_init_string(cfg); // send initiation string to modem
   cfg->line_data.fd=ip_accept(cfg->line_data.sfd);
   if(cfg->line_data.fd > -1) {
     cfg->line_data.valid_conn=TRUE;
