--- /src/smtp.lua	2015-09-24 16:31:14.587402905 +0300

@@ -90,7 +90,7 @@
 end
 
 function metat.__index:auth(user, password, ext)
-    if not user or not password then return 1 end
+    if not user or user == "" or password == "" or not password then return 1 end
     if string.find(ext, "AUTH[^\n]+LOGIN") then
         return self:login(user, password)
     elseif string.find(ext, "AUTH[^\n]+PLAIN") then
