--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -2506,8 +2506,9 @@
 	} else if (os_strcmp(buf, "skip_inactivity_poll") == 0) {
 		bss->skip_inactivity_poll = atoi(pos);
 	} else if (os_strcmp(buf, "country_code") == 0) {
-		if (pos[0] < 'A' || pos[0] > 'Z' ||
-		    pos[1] < 'A' || pos[1] > 'Z') {
+		if ((pos[0] < 'A' || pos[0] > 'Z' ||
+		    pos[1] < 'A' || pos[1] > 'Z') &&
+		   !(pos[0] == '0' && pos[1] == '0')){
 			wpa_printf(MSG_ERROR,
 				   "Line %d: Invalid country_code '%s'",
 				   line, pos);

