Index: linux-3.18.44/drivers/net/usb/asix_devices.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/asix_devices.c
+++ linux-3.18.44/drivers/net/usb/asix_devices.c
@@ -195,6 +195,7 @@ static const struct net_device_ops ax881
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address 	= eth_mac_addr,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_do_ioctl		= asix_ioctl,
Index: linux-3.18.44/drivers/net/usb/ax88179_178a.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/ax88179_178a.c
+++ linux-3.18.44/drivers/net/usb/ax88179_178a.c
@@ -961,6 +961,7 @@ static const struct net_device_ops ax881
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= ax88179_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address	= ax88179_set_mac_addr,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_do_ioctl		= ax88179_ioctl,
Index: linux-3.18.44/drivers/net/usb/dm9601.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/dm9601.c
+++ linux-3.18.44/drivers/net/usb/dm9601.c
@@ -344,6 +344,7 @@ static const struct net_device_ops dm960
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_do_ioctl 		= dm9601_ioctl,
 	.ndo_set_rx_mode	= dm9601_set_multicast,
Index: linux-3.18.44/drivers/net/usb/int51x1.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/int51x1.c
+++ linux-3.18.44/drivers/net/usb/int51x1.c
@@ -144,6 +144,7 @@ static const struct net_device_ops int51
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address	= eth_mac_addr,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_set_rx_mode	= int51x1_set_multicast,
Index: linux-3.18.44/drivers/net/usb/mcs7830.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/mcs7830.c
+++ linux-3.18.44/drivers/net/usb/mcs7830.c
@@ -476,6 +476,7 @@ static const struct net_device_ops mcs78
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_do_ioctl 		= mcs7830_ioctl,
 	.ndo_set_rx_mode	= mcs7830_set_multicast,
Index: linux-3.18.44/drivers/net/usb/qmi_wwan.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/qmi_wwan.c
+++ linux-3.18.44/drivers/net/usb/qmi_wwan.c
@@ -241,6 +241,7 @@ static const struct net_device_ops qmi_w
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address	= qmi_wwan_mac_addr,
 	.ndo_validate_addr	= eth_validate_addr,
 };
Index: linux-3.18.44/drivers/net/usb/rndis_host.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/rndis_host.c
+++ linux-3.18.44/drivers/net/usb/rndis_host.c
@@ -292,6 +292,7 @@ static const struct net_device_ops rndis
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_set_mac_address 	= eth_mac_addr,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_validate_addr	= eth_validate_addr,
 };
 
Index: linux-3.18.44/drivers/net/usb/sierra_net.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/sierra_net.c
+++ linux-3.18.44/drivers/net/usb/sierra_net.c
@@ -174,6 +174,7 @@ static const struct net_device_ops sierr
 	.ndo_start_xmit         = usbnet_start_xmit,
 	.ndo_tx_timeout         = usbnet_tx_timeout,
 	.ndo_change_mtu         = sierra_net_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address    = eth_mac_addr,
 	.ndo_validate_addr      = eth_validate_addr,
 };
Index: linux-3.18.44/drivers/net/usb/smsc75xx.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/smsc75xx.c
+++ linux-3.18.44/drivers/net/usb/smsc75xx.c
@@ -1373,6 +1373,7 @@ static const struct net_device_ops smsc7
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= smsc75xx_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address 	= eth_mac_addr,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_do_ioctl 		= smsc75xx_ioctl,
Index: linux-3.18.44/drivers/net/usb/smsc95xx.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/smsc95xx.c
+++ linux-3.18.44/drivers/net/usb/smsc95xx.c
@@ -1092,6 +1092,7 @@ static const struct net_device_ops smsc9
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address 	= eth_mac_addr,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_do_ioctl 		= smsc95xx_ioctl,
Index: linux-3.18.44/drivers/net/usb/sr9700.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/sr9700.c
+++ linux-3.18.44/drivers/net/usb/sr9700.c
@@ -308,6 +308,7 @@ static const struct net_device_ops sr970
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_do_ioctl		= sr9700_ioctl,
 	.ndo_set_rx_mode	= sr9700_set_multicast,
Index: linux-3.18.44/drivers/net/usb/sr9800.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/sr9800.c
+++ linux-3.18.44/drivers/net/usb/sr9800.c
@@ -683,6 +683,7 @@ static const struct net_device_ops sr980
 	.ndo_start_xmit		= usbnet_start_xmit,
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address	= sr_set_mac_address,
 	.ndo_validate_addr	= eth_validate_addr,
 	.ndo_do_ioctl		= sr_ioctl,
Index: linux-3.18.44/drivers/net/usb/usbnet.c
===================================================================
--- linux-3.18.44.orig/drivers/net/usb/usbnet.c
+++ linux-3.18.44/drivers/net/usb/usbnet.c
@@ -317,6 +317,7 @@ static void __usbnet_status_stop_force(s
  */
 void usbnet_skb_return (struct usbnet *dev, struct sk_buff *skb)
 {
+	struct pcpu_sw_netstats *stats64 = this_cpu_ptr(dev->stats64);
 	int	status;
 
 	if (test_bit(EVENT_RX_PAUSED, &dev->flags)) {
@@ -325,8 +326,10 @@ void usbnet_skb_return (struct usbnet *d
 	}
 
 	skb->protocol = eth_type_trans (skb, dev->net);
-	dev->net->stats.rx_packets++;
-	dev->net->stats.rx_bytes += skb->len;
+	u64_stats_update_begin(&stats64->syncp);
+	stats64->rx_packets++;
+	stats64->rx_bytes += skb->len;
+	u64_stats_update_end(&stats64->syncp);
 
 	netif_dbg(dev, rx_status, dev->net, "< rx, len %zu, type 0x%x\n",
 		  skb->len + sizeof (struct ethhdr), skb->protocol);
@@ -962,6 +965,37 @@ int usbnet_set_settings (struct net_devi
 }
 EXPORT_SYMBOL_GPL(usbnet_set_settings);
 
+void usbnet_get_stats64(struct net_device *net, struct rtnl_link_stats64 *stats)
+{
+	struct usbnet *dev = netdev_priv(net);
+	unsigned int start;
+	int cpu;
+
+	netdev_stats_to_stats64(stats, &net->stats);
+
+	for_each_possible_cpu(cpu) {
+		struct pcpu_sw_netstats *stats64;
+		u64 rx_packets, rx_bytes;
+		u64 tx_packets, tx_bytes;
+
+		stats64 = per_cpu_ptr(dev->stats64, cpu);
+
+		do {
+			start = u64_stats_fetch_begin_irq(&stats64->syncp);
+			rx_packets = stats64->rx_packets;
+			rx_bytes = stats64->rx_bytes;
+			tx_packets = stats64->tx_packets;
+			tx_bytes = stats64->tx_bytes;
+		} while (u64_stats_fetch_retry_irq(&stats64->syncp, start));
+
+		stats->rx_packets += rx_packets;
+		stats->rx_bytes += rx_bytes;
+		stats->tx_packets += tx_packets;
+		stats->tx_bytes += tx_bytes;
+	}
+}
+EXPORT_SYMBOL_GPL(usbnet_get_stats64);
+
 u32 usbnet_get_link (struct net_device *net)
 {
 	struct usbnet *dev = netdev_priv(net);
@@ -1193,8 +1227,12 @@ static void tx_complete (struct urb *urb
 	struct usbnet		*dev = entry->dev;
 
 	if (urb->status == 0) {
-		dev->net->stats.tx_packets += entry->packets;
-		dev->net->stats.tx_bytes += entry->length;
+		struct pcpu_sw_netstats *stats64 = this_cpu_ptr(dev->stats64);
+
+		u64_stats_update_begin(&stats64->syncp);
+		stats64->tx_packets += entry->packets;
+		stats64->tx_bytes += entry->length;
+		u64_stats_update_end(&stats64->syncp);
 	} else {
 		dev->net->stats.tx_errors++;
 
@@ -1551,6 +1589,7 @@ void usbnet_disconnect (struct usb_inter
 	usb_free_urb(dev->interrupt);
 	kfree(dev->padding_pkt);
 
+	free_percpu(dev->stats64);
 	free_netdev(net);
 }
 EXPORT_SYMBOL_GPL(usbnet_disconnect);
@@ -1562,6 +1601,7 @@ static const struct net_device_ops usbne
 	.ndo_tx_timeout		= usbnet_tx_timeout,
 	.ndo_set_rx_mode	= usbnet_set_rx_mode,
 	.ndo_change_mtu		= usbnet_change_mtu,
+	.ndo_get_stats64        = usbnet_get_stats64,
 	.ndo_set_mac_address 	= eth_mac_addr,
 	.ndo_validate_addr	= eth_validate_addr,
 };
@@ -1623,6 +1663,11 @@ usbnet_probe (struct usb_interface *udev
 	dev->intf = udev;
 	dev->driver_info = info;
 	dev->driver_name = name;
+
+	dev->stats64 = netdev_alloc_pcpu_stats(struct pcpu_sw_netstats);
+	if (!dev->stats64)
+		goto out0;
+
 	dev->msg_enable = netif_msg_init (msg_level, NETIF_MSG_DRV
 				| NETIF_MSG_PROBE | NETIF_MSG_LINK);
 	init_waitqueue_head(&dev->wait);
@@ -1775,6 +1820,8 @@ out1:
 	 */
 	cancel_work_sync(&dev->kevent);
 	del_timer_sync(&dev->delay);
+	free_percpu(dev->stats64);
+out0:
 	free_netdev(net);
 out:
 	return status;
Index: linux-3.18.44/include/linux/usb/usbnet.h
===================================================================
--- linux-3.18.44.orig/include/linux/usb/usbnet.h
+++ linux-3.18.44/include/linux/usb/usbnet.h
@@ -64,6 +64,8 @@ struct usbnet {
 	struct usb_anchor	deferred;
 	struct tasklet_struct	bh;
 
+	struct pcpu_sw_netstats __percpu *stats64;
+
 	struct work_struct	kevent;
 	unsigned long		flags;
 #		define EVENT_TX_HALT	0
@@ -278,5 +280,7 @@ extern int usbnet_status_start(struct us
 extern void usbnet_status_stop(struct usbnet *dev);
 
 extern void usbnet_update_max_qlen(struct usbnet *dev);
+extern void usbnet_get_stats64(struct net_device *dev,
+                              struct rtnl_link_stats64 *stats);
 
 #endif /* __LINUX_USB_USBNET_H */
