Index: linux-3.18.44/drivers/mtd/spi-nor/spi-nor.c
===================================================================
--- linux-3.18.44.orig/drivers/mtd/spi-nor/spi-nor.c	2018-04-24 15:37:52.923013165 +0300
+++ linux-3.18.44/drivers/mtd/spi-nor/spi-nor.c	2018-04-24 15:42:08.463018249 +0300
@@ -512,6 +512,9 @@
 	{ "gd25q64", INFO(0xc84017, 0, 64 * 1024, 128, SECT_4K) },
 	{ "gd25q128", INFO(0xc84018, 0, 64 * 1024, 256, SECT_4K) },
 
+	/* Zbit */
+	{ "zb25vq128", INFO(0xC84018, 0, 64 * 1024,  64, SECT_4K) },
+
 	/* Intel/Numonyx -- xxxs33b */
 	{ "160s33b",  INFO(0x898911, 0, 64 * 1024,  32, 0) },
 	{ "320s33b",  INFO(0x898912, 0, 64 * 1024,  64, 0) },
