--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c	
@@ -649,9 +649,13 @@
 
 	/* XTX */
 	{ "xt25f128a", INFO(0x207018, 0, 64 * 1024, 256, SECT_4K) },	
+	{ "xt25f128b", INFO(0x0B4018, 0, 64 * 1024, 256, SECT_4K) },	
 
 	{ },
 };
+
+static const struct spi_device_id spi_nor_unknown_id = 
+	{ "undefined",  INFO(0x0, 0, 64 * 1024, 256, SECT_4K) };
 
 static const struct spi_device_id *spi_nor_read_id(struct spi_nor *nor)
 {
@@ -681,8 +685,9 @@
 				return &spi_nor_ids[tmp];
 		}
 	}
-	dev_err(nor->dev, "unrecognized JEDEC id %06x\n", jedec);
-	return ERR_PTR(-ENODEV);
+	return &spi_nor_unknown_id; // for Teltonikia RUT devices
+	//dev_err(nor->dev, "unrecognized JEDEC id %06x\n", jedec);
+	//return ERR_PTR(-ENODEV);
 }
 
 static int spi_nor_read(struct mtd_info *mtd, loff_t from, size_t len,
