--- a/include/linux/usb/cdc_ncm.h
+++ b/include/linux/usb/cdc_ncm.h
@@ -49,12 +49,12 @@
 #define USB_CDC_NCM_NDP16_LENGTH_MIN		0x10
 
 /* Maximum NTB length */
-#define	CDC_NCM_NTB_MAX_SIZE_TX			32768	/* bytes */
-#define	CDC_NCM_NTB_MAX_SIZE_RX			32768	/* bytes */
+#define	CDC_NCM_NTB_MAX_SIZE_TX			8192	/* bytes */
+#define	CDC_NCM_NTB_MAX_SIZE_RX			4096	/* bytes */
 
 /* Initial NTB length */
-#define	CDC_NCM_NTB_DEF_SIZE_TX			16384	/* bytes */
-#define	CDC_NCM_NTB_DEF_SIZE_RX			16384	/* bytes */
+#define	CDC_NCM_NTB_DEF_SIZE_TX			8192	/* bytes */
+#define	CDC_NCM_NTB_DEF_SIZE_RX			4096	/* bytes */
 
 /* Minimum value for MaxDatagramSize, ch. 6.2.9 */
 #define	CDC_NCM_MIN_DATAGRAM_SIZE		1514	/* bytes */
@@ -65,7 +65,7 @@
 #define	CDC_NCM_MIN_TX_PKT			512	/* bytes */
 
 /* Default value for MaxDatagramSize */
-#define	CDC_NCM_MAX_DATAGRAM_SIZE		8192	/* bytes */
+#define	CDC_NCM_MAX_DATAGRAM_SIZE		2048	/* bytes */
 
 /*
  * Maximum amount of datagrams in NCM Datagram Pointer Table, not counting
