Index: libiwinfo-2014-10-27.1/iwinfo_lib.c
===================================================================
--- libiwinfo-2014-10-27.1.orig/iwinfo_lib.c
+++ libiwinfo-2014-10-27.1/iwinfo_lib.c
@@ -176,7 +176,6 @@ const struct iwinfo_iso3166_label IWINFO
 	{ 0x4A45 /* JE */, "Jersey" },
 	{ 0x4A4D /* JM */, "Jamaica" },
 	{ 0x4A4F /* JO */, "Jordan" },
-	{ 0x4A50 /* JP */, "Japan" },
 	{ 0x4B45 /* KE */, "Kenya" },
 	{ 0x4B47 /* KG */, "Kyrgyzstan" },
 	{ 0x4B48 /* KH */, "Cambodia" },
